exporter_test.go

```diff
@@ -173,3 +168,211 @@ func TestEscapeMetricName(t *testing.T) {
 		}
 	}
 }
+
+// TestTtlExpiration validates expiration of time series.
+// foobar metric without mapping should expire with default ttl of 1s
+// bazqux metric should expire with ttl of 2s
+func TestTtlExpiration(t *testing.T) {
+	config := `
+defaults:
+  ttl: 1s
+mappings:
+- match: bazqux.*
+  name: bazqux
+  labels:
+    first: baz
+    second: qux
+    third: $1
+  ttl: 2s
+`
+
+	bazquxLabels := prometheus.Labels{
+		"third":  "main",
+		"first":  "baz",
+		"second": "qux",
+	}
+
+	testMapper := &mapper.MetricMapper{}
+	err := testMapper.InitFromYAMLString(config)
+	if err != nil {
+		t.Fatalf("Config load error: %s %s", config, err)
+	}
+
+	ex := NewExporter(testMapper)
+	for _, l := range []statsDPacketHandler{&StatsDUDPListener{}, &mockStatsDTCPListener{}} {
+		events := make(chan Events, 2)
+		fatal := make(chan error, 1) // t.Fatal must not be called in goroutines (SA2002)
+		stop := make(chan bool, 1)
+
+		l.handlePacket([]byte("foobar:200|g"), events)
+		l.handlePacket([]byte("bazqux.main:42|ms"), events)
+
+		// Close channel to signify we are done with the listener after a short period.
+		go func() {
+			defer close(events)
+
+			time.Sleep(time.Millisecond * 100)
```

Comment:
Could we test this without _actually_ sleeping (and depending on actual wall clocks and such)? I found an [article](https://smartystreets.com/blog/2015/09/go-testing-part-5-testing-with-time) that makes some good suggestions.

Reply:
Done